data_safe_haven/config/context_manager.py

```diff
@@ -91,50 +90,49 @@ def available(self) -> list[str]:
     def update(
         self,
         *,
-        admin_group_id: str | None = None,
-        location: str | None = None,
+        admin_group_name: str | None = None,
         name: str | None = None,
         subscription_name: str | None = None,
     ) -> None:
         context = self.assert_context()
 
-        if admin_group_id:
+        if admin_group_name:
             self.logger.debug(
-                f"Updating '[green]{admin_group_id}[/]' to '{admin_group_id}'."
+                f"Updating admin group name from '{context.admin_group_name}' to '[green]{admin_group_name}[/]'."
             )
-            context.admin_group_id = admin_group_id
-        if location:
-            self.logger.debug(f"Updating '[green]{location}[/]' to '{location}'.")
-            context.location = location
+            context.admin_group_name = admin_group_name
         if name:
-            self.logger.debug(f"Updating '[green]{name}[/]' to '{name}'.")
+            self.logger.debug(
+                f"Updating name from '{context.name}' to '[green]{name}[/]'."
+            )
             context.name = name
         if subscription_name:
             self.logger.debug(
-                f"Updating '[green]{subscription_name}[/]' to '{subscription_name}'."
+                f"Updating subscription name from '{context.subscription_name}' to '[green]{subscription_name}[/]'."
             )
             context.subscription_name = subscription_name
 
     def add(
         self,
         *,
-        key: str,
+        admin_group_name: str,
         name: str,
-        admin_group_id: str,
-        location: str,
         subscription_name: str,
     ) -> None:
         # Ensure context is not already present
+        key = json_safe(name)
```

Comment:
I don't like taking user input and then modifying it opaquely where two different inputs can give the same output, like here.

If we want to only have one, I think it would be better to have a validator ensuring that `name == json_safe(name)`.

Reply:
Sure - that makes sense. I think we could expand the definition of JSON-safe to include other characters (underscores and hyphens would make sense - perhaps others too?).

Reply:
This is actually more involved than I thought at first. I suggest leaving this PR as-is and discussing further on #1965.

Reply:
We can't easily undo the changes here?

Reply:
Done in 0f5e6c4